pipelines/controller/src/main/java/org/openmrs/analytics/ApiController.java

```diff
@@ -64,6 +64,15 @@ public ProgressStats getStatus() {
     return progressStats;
   }
 
+  @PostMapping("/tables")
+  public void createResourceTables() {
+    if (pipelineManager.isRunning()) {
+      throw new IllegalStateException("Another pipeline is running!");
```

Comment:
nit: This should probably be "cannot create tables while a pipeline is running" or something like that. Also it would be great if we can expose the error in the UI.

Regarding concurrency, do we need to care about a situation when two requests come from the UI? It is an edge case and I think we handle it gracefully at least right now but maybe add a comment about its safety if you don't want to introduce another lock.

Reply:
Sure, will take care of the first comment. Regarding second, when will be the case where we might experience unwanted behaviour even if application receives multiple requests?

Reply:
I don't think it is a big deal right now but making up examples: Two requests come at around the same time but one picks DWH files at T1 and the other at T2. Then depending on race conditions, one create Patient canonical tables for T1 while the other wins creating Observation canonical tables for T2. So we may end up with Patient and Observation canonical tables that are inconsistent with each other.

Reply:
But, this will happen when we allow multiple instances of pipeline to run. If we are not allowing any further run when the pipeline is already running then both requests should land on the latest snapshots, right?

Reply:
For the above example, we don't need two instances of the pipeline to be running at the same time. So first request picks DWH of T1 because the pipeline for T2 is not done yet. Then the pipeline is done and second request picks T2 (if we want to be pedantic, the pipeline may even start after the first request and finish quickly). With the `if (pipelineManager.isRunning())` condition above, this scenario is very unlikely and that's why I think it is not a big deal. But in principle, you can have two threads changing the same resource (i.e., the canonical tables on the thrift-server) without synchronization between them.

Reply:
BTW, please do not hold this PR on this edge case; feel free to merge if other comments are addressed.

Reply:
sure thanks